src/coreComponents/physicsSolvers/SolverBase.cpp

```diff
@@ -458,162 +562,223 @@ real64 SolverBase::nonlinearImplicitStep( real64 const & time_n,
-      debugOutputSystem( time_n, cycleNumber, newtonIter, m_matrix, m_rhs );
+        // Output the linear system matrix/rhs for debugging purposes
+        debugOutputSystem( time_n, cycleNumber, newtonIter, m_matrix, m_rhs );
 
-      // Solve the linear system
-      solveSystem( m_dofManager, m_matrix, m_rhs, m_solution );
+        // Solve the linear system
+        solveSystem( m_dofManager, m_matrix, m_rhs, m_solution );
 
-      // Output the linear system solution for debugging purposes
-      debugOutputSolution( time_n, cycleNumber, newtonIter, m_solution );
+        // Output the linear system solution for debugging purposes
+        debugOutputSolution( time_n, cycleNumber, newtonIter, m_solution );
 
-      scaleFactor = scalingForSystemSolution( domain, m_dofManager, m_solution.values() );
+        scaleFactor = scalingForSystemSolution( domain, m_dofManager, m_solution.values() );
 
-      if( !checkSystemSolution( domain, m_dofManager, m_solution.values(), scaleFactor ) )
-      {
-        // TODO try chopping (similar to line search)
-        GEOSX_LOG_RANK_0( "    Solution check failed. Newton loop terminated." );
-        break;
-      }
+        if( !checkSystemSolution( domain, m_dofManager, m_solution.values(), scaleFactor ) )
+        {
+          // TODO try chopping (similar to line search)
+          GEOSX_LOG_RANK_0( "    Solution check failed. Newton loop terminated." );
+          break;
+        }
 
-      // apply the system solution to the fields/variables
-      applySystemSolution( m_dofManager, m_solution.values(), scaleFactor, domain );
+        // apply the system solution to the fields/variables
+        applySystemSolution( m_dofManager, m_solution.values(), scaleFactor, domain );
 
-      // update non-primary variables (constitutive models)
-      updateState( domain );
+        // update non-primary variables (constitutive models)
+        updateState( domain );
 
-      lastResidual = residualNorm;
-    }
+        lastResidual = residualNorm;
+      } // end of Newton's loop
 
-    if( isConverged )
+      if( isNewtonConverged )
+      {
+        isConfigurationLoopConverged = updateConfiguration( domain );
+
+        if( isConfigurationLoopConverged )
+        {
+          break; // get out of configuration loop coz everything converged.
+        }
+        else
+        {
+          GEOSX_LOG_LEVEL_RANK_0( 1, "  --------Configuration did not converge. Testing new configuration." );
+        }
+      }
+      else if( !attemptedSimplestConfiguration )
+      {
+        resetStateToBeginningOfStep( domain );
+        bool const breakLoop = setSimplestConfigurationState( domain );
+        attemptedSimplestConfiguration = true;
+        if( breakLoop )
+        {
+          break;
+        }
+      }
+      else
+      {
+        break; // get out of configuration loop and cut the time-step if you can.
+      }
+
+    }  // end of configuration loop
```

Comment:
We can imagine to have a policy to make convergence easier, but I think that this policy could be `Solver` based and not imposed to all solvers.

Maybe a code like
```c++
if( isNewtonConverged ) simplify( domain );
```
which lets all the solvers implement their own versions of `simplify` could be a good balance.
If this removes `resetStateToBeginningOfStep`, `setSimplestConfigurationState` or `updateConfiguration` from the `public` interface of `SolverBase` which is already bloat, that could be great.

Reply:
I guess something of that sort would clean up `SolverBase` which is certainly needed. I am not sure `simplify` is the best name for it though. Does anyone have another suggestion?